cvat-core/src/frames.ts

```diff
@@ -157,233 +157,173 @@ export class FrameData {
-        };
+Object.defineProperty(FrameData.prototype.data, 'implementation', {
+    value(this: FrameData, onServerRequest) {
+        return new Promise<{
+            renderWidth: number;
+            renderHeight: number;
+            imageData: ImageBitmap | Blob;
+        } | Blob>((resolve, reject) => {
+            const {
+                provider, chunkSize, stopFrame, decodeForward, forwardStep, decodedBlocksCacheSize,
+            } = frameDataCache[this.jobID];
+
+            const requestId = +_.uniqueId();
+            const chunkNumber = Math.floor(this.number / chunkSize);
+            const frame = provider.frame(this.number);
+
+            function findTheNextNotDecodedChunk(searchFrom: number): number {
+                let firstFrameInNextChunk = searchFrom + forwardStep;
+                let nextChunkNumber = Math.floor(firstFrameInNextChunk / chunkSize);
+                while (nextChunkNumber === chunkNumber) {
+                    firstFrameInNextChunk += forwardStep;
+                    nextChunkNumber = Math.floor(firstFrameInNextChunk / chunkSize);
+                }
 
-        if (this._data) {
-            resolve(this._data);
-            return;
-        }
+                if (provider.isChunkCached(nextChunkNumber)) {
+                    return findTheNextNotDecodedChunk(firstFrameInNextChunk);
+                }
 
-        const { provider } = frameDataCache[this.jid];
-        const { chunkSize } = frameDataCache[this.jid];
-        const start = parseInt(this.number / chunkSize, 10) * chunkSize;
-        const stop = Math.min(this.stopFrame, (parseInt(this.number / chunkSize, 10) + 1) * chunkSize - 1);
-        const chunkNumber = Math.floor(this.number / chunkSize);
-
-        const onDecodeAll = async (frameNumber) => {
-            if (
-                frameDataCache[this.jid].activeChunkRequest &&
-                chunkNumber === frameDataCache[this.jid].activeChunkRequest.chunkNumber
-            ) {
-                const callbackArray = frameDataCache[this.jid].activeChunkRequest.callbacks;
-                for (let i = callbackArray.length - 1; i >= 0; --i) {
-                    if (callbackArray[i].frameNumber === frameNumber) {
-                        const callback = callbackArray[i];
-                        callbackArray.splice(i, 1);
-                        callback.resolve(await provider.frame(callback.frameNumber));
+                return nextChunkNumber;
+            }
+
+            if (frame) {
+                if (decodeForward && decodedBlocksCacheSize > 1 && !frameDataCache[this.jobID].activeChunkRequest) {
+                    const nextChunkNumber = findTheNextNotDecodedChunk(this.number);
+                    const predecodeChunksMax = Math.floor(decodedBlocksCacheSize / 2);
+                    if (nextChunkNumber * chunkSize <= stopFrame &&
+                        nextChunkNumber <= chunkNumber + predecodeChunksMax) {
+                        provider.cleanup(1);
```

Comment:
Isn't `1` default argument for `cleanup`?

Reply:
Actually yes, but explicitly specifying it here is not extra